File: haskell/hoascbv.hs
{-# LANGUAGE BangPatterns #-}

module Main where

data Value
  = VVar !Int
  | VApp !Value !Value
  | VLam (Value -> Value)

data Tm
  = Var !Int
  | App !Tm !Tm
  | Lam !Tm

ap :: Value -> Value -> Value
ap t u =
  case t of
    VLam f -> f u
    _ -> VApp t u

ap2 :: Value -> Value -> Value -> Value
ap2 t u v = ap (ap t u) v

n2 :: Value
n2 = VLam (\s -> VLam (\z -> ap s (ap s z)))

n5 :: Value
n5 = VLam (\s -> VLam (\z -> ap s (ap s (ap s (ap s (ap s z))))))

mul :: Value
mul =
  VLam
    ( \a ->
        VLam
          ( \b ->
              VLam
                (\s -> VLam (\z -> ap (ap a (ap b s)) z))
          )
    )

suc :: Value -> Value
suc n = VLam (\s -> VLam (\z -> ap s (ap2 n s z)))

n10 :: Value
n10 = ap2 mul n2 n5

n20 :: Value
n20 = ap2 mul n2 n10

n21 :: Value
n21 = suc n20

n22 :: Value
n22 = suc n21

leaf :: Value
leaf = VLam (\l -> VLam (\_ -> l))

node :: Value
node =
  VLam
    ( \t1 ->
        VLam
          ( \t2 ->
              VLam
                (\_ -> VLam (\n -> ap2 n t1 t2))
          )
    )

fullTree :: Value
fullTree = VLam (\n -> ap2 n (VLam (\t -> ap2 node t t)) leaf)

quote :: Int -> Value -> Tm
quote l v =
  case v of
    VVar x -> Var (l - x - 1)
    VLam t -> Lam (quote (l + 1) (t (VVar l)))
    VApp t u -> App (quote l t) (quote l u)

quote0 :: Value -> Tm
quote0 = quote 0

force :: Tm -> Int
force t =
  case t of
    Var _ -> 0
    App t1 t2 -> force t1 + force t2
    Lam body -> 1 + force body

main :: IO ()
main = do
  let !lamCount = force (quote0 (ap fullTree n22))
  if lamCount /= 16777214
    then error ("FAIL: expected 16777214, got " ++ show lamCount)
    else pure ()

File: rust/hoascbv.rs
#![allow(dead_code)]

use std::rc::Rc;

enum Term {
    Var(i64),
    App(Rc<Term>, Rc<Term>),
    Lam(Rc<Term>),
}

#[derive(Clone)]
enum Value {
    VVar(i64),
    VApp(Rc<Value>, Rc<Value>),
    VLam(Rc<dyn Fn(Value) -> Value>),
}

fn vlam<F>(f: F) -> Value
where
    F: Fn(Value) -> Value + 'static,
{
    Value::VLam(Rc::new(f))
}

fn ap(t: Value, u: Value) -> Value {
    match t {
        Value::VLam(f) => f(u),
        t => Value::VApp(Rc::new(t), Rc::new(u)),
    }
}

fn ap2(t: Value, u: Value, v: Value) -> Value {
    ap(ap(t, u), v)
}

fn n2() -> Value {
    vlam(|s| vlam(move |z| ap(s.clone(), ap(s.clone(), z))))
}

fn n5() -> Value {
    vlam(|s| {
        vlam(move |z| ap(s.clone(), ap(s.clone(), ap(s.clone(), ap(s.clone(), ap(s.clone(), z))))))
    })
}

fn mul() -> Value {
    vlam(|a| {
        vlam(move |b| {
            let a1 = a.clone();
            vlam(move |s| {
                let a2 = a1.clone();
                let b1 = b.clone();
                vlam(move |z| ap(ap(a2.clone(), ap(b1.clone(), s.clone())), z))
            })
        })
    })
}

fn suc(n: Value) -> Value {
    vlam(move |s| {
        let n1 = n.clone();
        vlam(move |z| ap(s.clone(), ap2(n1.clone(), s.clone(), z)))
    })
}

fn leaf() -> Value {
    vlam(|l| vlam(move |_n| l.clone()))
}

fn node() -> Value {
    vlam(|t1| {
        vlam(move |t2| {
            let t1a = t1.clone();
            vlam(move |_l| {
                let t1b = t1a.clone();
                let t2b = t2.clone();
                vlam(move |n| ap2(n, t1b.clone(), t2b.clone()))
            })
        })
    })
}

fn full_tree() -> Value {
    vlam(|n| ap2(n, vlam(|t| ap2(node(), t.clone(), t)), leaf()))
}

fn quote(level: i64, v: &Value) -> Term {
    match v {
        Value::VVar(x) => Term::Var(level - x - 1),
        Value::VApp(t, u) => Term::App(Rc::new(quote(level, t)), Rc::new(quote(level, u))),
        Value::VLam(f) => {
            let body = f(Value::VVar(level));
            Term::Lam(Rc::new(quote(level + 1, &body)))
        }
    }
}

fn quote0(v: &Value) -> Term {
    quote(0, v)
}

fn force(t: &Term) -> i64 {
    match t {
        Term::Var(_) => 0,
        Term::App(t1, t2) => force(t1) + force(t2),
        Term::Lam(body) => 1 + force(body),
    }
}

fn main() {
    let n10 = ap2(mul(), n2(), n5());
    let n20 = ap2(mul(), n2(), n10);
    let n21 = suc(n20);
    let n22 = suc(n21);

    let tree8m = ap(full_tree(), n22);
    let quoted = quote0(&tree8m);
    let lam_count = force(&quoted);

    if lam_count != 16_777_214 {
        eprintln!("FAIL: expected 16777214, got {}", lam_count);
        std::process::exit(1);
    }
}
